pkg/node/manager/manager.go

```diff
@@ -0,0 +1,347 @@
+	switch newSource {
+	// the local node always takes precedence
+	case node.FromLocalNode:
+		return true
+
+	// agent local updates can overwrite everything except for the local
+	// node
+	case node.FromAgentLocal:
+		return oldSource != node.FromLocalNode
+
+	// kvstore updates can overwrite everything except agent local updates and local node
+	case node.FromKVStore:
+		return oldSource != node.FromAgentLocal && oldSource != node.FromLocalNode
+
+	// kubernetes updates can only overwrite kubernetes nodes
+	case node.FromKubernetes:
+		return oldSource == node.FromKubernetes
+
+	default:
+		return false
+	}
+}
+
+// NodeUpdated is called after the information of a node has been updated. The
+// node in the manager is added or updated if the source is allowed to update
+// the node. If an update or addition has occured, NodeUpdate() of the datapath
+// interface is invoked.
+func (m *Manager) NodeUpdated(n node.Node) {
+
+	log.Debugf("Received node update event from %s: %#v", n.Source, n)
+
+	nodeIdentity := n.Identity()
+
+	m.mutex.Lock()
+	entry, oldNodeExists := m.nodes[nodeIdentity]
+	if oldNodeExists {
+		m.metricEventsReceived.WithLabelValues("update").Inc()
+
+		if !overwriteAllowed(entry.node.Source, n.Source) {
+			m.mutex.Unlock()
+			return
+		}
+
+		entry.mutex.Lock()
+		m.mutex.Unlock()
+		oldNode := entry.node
+		entry.node = n
+		m.datapath.NodeUpdate(oldNode, entry.node)
+		entry.mutex.Unlock()
+	} else {
+		m.metricEventsReceived.WithLabelValues("add").Inc()
+		m.metricNumNodes.Inc()
+
+		entry = &nodeEntry{node: n}
+		entry.mutex.Lock()
+		m.nodes[nodeIdentity] = entry
+		m.mutex.Unlock()
+		m.datapath.NodeAdd(entry.node)
+		entry.mutex.Unlock()
+	}
+}
+
+// NodeDeleted is called after a node has been deleted. It removes the node
+// from the manager if the node is still owned by the source of which the event
+// orgins from. If the node was removed, NodeDelete() is invoked of the
+// datapath interface.
+func (m *Manager) NodeDeleted(n node.Node) {
+	m.metricEventsReceived.WithLabelValues("delete").Inc()
+
+	log.Debugf("Received node delete event from %s", n.Source)
+
+	nodeIdentity := n.Identity()
+
+	m.mutex.Lock()
+	entry, oldNodeExists := m.nodes[nodeIdentity]
+	if !oldNodeExists {
+		m.mutex.Unlock()
+		return
+	}
+
+	if n.Source != entry.node.Source {
+		log.Debugf("Ignoring delete event of node %s from source %s. The node is owned by %s",
+			n.Name, n.Source, entry.node.Source)
+		m.mutex.Unlock()
+		return
+	}
+
+	m.metricNumNodes.Dec()
+
+	entry.mutex.Lock()
+	delete(m.nodes, nodeIdentity)
+	m.mutex.Unlock()
+	m.datapath.NodeDelete(n)
+	entry.mutex.Unlock()
+}
+
+// GetNodes returns a copy of all of the nodes as a map from Identity to Node.
+func (m *Manager) GetNodes() map[node.Identity]node.Node {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	nodes := make(map[node.Identity]node.Node)
+	for nodeIdentity, entry := range m.nodes {
+		entry.mutex.Lock()
```

Comment:
What is this mutex protecting? I looked through all code and doesn't seem to do anything relevant as its access is already being protected by `m.mutex`

Reply:
The documentation section of this mutex says:
```
        // mutex serves two purposes:
        // 1. Serialize any direct access to the node field in this entry.
        // 2. Serialize all calls do the datapath layer for a particular node.
        //
        // See description of Manager.mutex for more details
        //
        // If both the nodeEntry.mutex and Manager.mutex must be held, then the
        // Manager.mutex must *always* be acquired first.
        mutex lock.Mutex
```

It fulfills point 1 for consistency. It is not strictly required as we don't unlock Manager.mutex in this case but it is good to be consistent here and make sure that the documented locking rules are always followed.